# Modelado de Ecuaciones Estructurales (SEM): Caso Práctico y Aplicación

## Introducción a la Sesión

En esta clase analizaremos cómo modelar constructos latentes (como confianza o calidad del servicio) a partir de variables observables, empleando SEM (Structural Equation Modeling) con la librería semopy. Trabajaremos con datos sintéticos, normalizaremos las variables antes del modelado, y veremos cómo calcular los valores individuales de los factores latentes utilizando las cargas factoriales estimadas por el modelo.

### 1. Planteamiento del caso y modelo

#### Escenario:
La empresa ficticia EcomInsights busca entender cómo influyen confianza en la plataforma y calidad del servicio (constructos latentes) sobre la satisfacción del cliente. Además, se estudia si el ingreso del hogar tiene algún impacto relevante.

#### Variables:

- Latentes:
   - Confianza en la plataforma: Uso, Seguridad, Recomendaciones
   - Calidad del servicio: Tiempo de respuesta, Resolución de problemas

- Observadas:
   - Ingreso del hogar, Satisfacción del cliente

---

## ¿Qué es el Modelado de Ecuaciones Estructurales (SEM)?

El **Modelado de Ecuaciones Estructurales (SEM)** es una técnica estadística que integra análisis de regresión, análisis de factores y análisis de trayectorias en un marco cohesivo. Es especialmente útil para:

- Investigar **constructos complejos** que no se pueden medir directamente (como la confianza o la satisfacción).
- Modelar **sistemas de relaciones** entre variables.
- Evaluar tanto los **efectos directos** como los **efectos indirectos** de las variables.

### ¿Cuándo usar SEM?

- Cuando necesitamos medir variables que no se observan directamente (variables **latentes**), como actitudes o percepciones.
- Cuando queremos estudiar sistemas de relaciones complejas entre varias variables y analizar cómo una variable puede influir en otra **a través de** una tercera variable (mediación).
- En investigaciones donde la **causalidad** y los efectos acumulativos son importantes.

### Aplicación en el Caso de EcomInsights

La empresa quiere saber cómo diferentes factores influyen en la satisfacción del cliente. SEM es ideal para este caso, ya que algunas variables clave como la **confianza en la plataforma** son latentes y no pueden medirse directamente.

---

## Pasos para Implementar SEM en la Empresa

### 1. **Entender el Modelo**

#### Variables Latentes
- **Confianza en la Plataforma**: No se puede observar directamente. Se mide a través de varios indicadores, como la frecuencia de uso, recomendaciones a otros usuarios, y la percepción de seguridad.
- **Calidad del Servicio**: Se mide mediante indicadores como la rapidez de entrega, la atención al cliente, y la resolución de problemas.

#### Variables Observables
- **Ingreso del Hogar**: Directamente observable.
- **Satisfacción del Cliente**: Una variable observada, pero influenciada por constructos latentes como la confianza y la calidad del servicio.

#### Relaciones Directas e Indirectas
- **Relación directa**: ¿Cómo afecta el ingreso directamente a la satisfacción del cliente?
- **Relación indirecta**: ¿Cómo influye el ingreso en la calidad del servicio, y cómo esto afecta la satisfacción?

### 2. **Especificar el Modelo de Medición**

Primero, definimos cómo las variables latentes (**confianza**, **calidad del servicio**) se relacionan con sus indicadores observados. Esto se hace mediante un modelo de **análisis factorial confirmatorio (CFA)**.

### 3. **Pasar al Modelo Estructural**

Después de establecer el modelo de medición, pasamos a la **parte estructural**, donde modelamos las relaciones entre las variables latentes y observables, incluyendo los efectos directos e indirectos.

---

## Elementos Clave en SEM

### 1. **Variables Exógenas y Endógenas**

- **Exógenas**: Variables independientes dentro del sistema, como el **ingreso del hogar**.
- **Endógenas**: Variables dependientes que están siendo explicadas dentro del modelo, como la **satisfacción del cliente**.

### 2. **Análisis de Covarianza**

SEM analiza las matrices de varianza/covarianza en lugar de los datos brutos. Esto permite modelar las relaciones entre las variables a un nivel más profundo.

### 3. **Estimación y Ajuste del Modelo**

El modelo es estimado utilizando **Máxima Verosimilitud (ML)**, que ajusta los parámetros maximizando la probabilidad de que los datos observados se ajusten al modelo teórico. Evaluamos el ajuste del modelo con métricas como el **AIC** o el **BIC** para asegurar que el modelo sea parsimonioso y explique bien los datos.

### 4. **Identificación del Modelo**

El modelo debe ser identificado correctamente. Si hay más incógnitas que datos observados, el modelo será **no identificado** y no podrá estimarse correctamente. Se pueden hacer ajustes para que el modelo esté **sobre-identificado**, lo que permite hacer inferencias sobre los parámetros.

---

## Ejemplo de Implementación Práctica

### Diagrama de Trayectoria

En el caso de **EcomInsights**, un modelo SEM podría tener este formato:

1. **Variables Latentes**:
   - Confianza en la plataforma → Indicadores como "Frecuencia de uso", "Percepción de seguridad".
   - Calidad del servicio → Indicadores como "Tiempo de respuesta", "Resolución de problemas".

2. **Variables Observables**:
   - Ingreso del Hogar
   - Satisfacción del Cliente

3. **Relaciones**:
   - Ingreso del hogar influye en la calidad del servicio (directo).
   - La calidad del servicio influye en la satisfacción del cliente (indirecto).
   - La confianza en la plataforma afecta directamente la satisfacción del cliente.

### ¿Qué buscamos con SEM?

1. **Confirmar hipótesis**: ¿La calidad del servicio actúa como mediador entre el ingreso y la satisfacción?
2. **Evaluar el ajuste del modelo**: ¿El modelo teórico refleja adecuadamente los datos reales?

---

## Conclusión

El **Modelado de Ecuaciones Estructurales (SEM)** es una herramienta esencial para entender relaciones complejas entre variables, permitiendo a las empresas como **EcomInsights** identificar no solo los efectos directos de sus estrategias en la satisfacción del cliente, sino también los efectos indirectos a través de constructos como la confianza y la calidad del servicio. SEM ayuda a las empresas a validar sus modelos teóricos y a mejorar su toma de decisiones con base en datos.

### Beneficios de SEM para EcomInsights:

- **Mejora de estrategias**: Identificación de los factores clave que influyen en la satisfacción.
- **Manejo de variables complejas**: Modela constructos difíciles de medir, como la confianza.
- **Optimización**: Evaluación de cómo la mejora en uno de los factores influye en los demás.

Este tipo de análisis permite decisiones estratégicas más informadas, ayudando a la empresa a optimizar la experiencia del cliente y a mejorar su propuesta de valor.


# Implementación de SEM en Python usando semopy

## Paso 1: Instalación de semopy
Instala la librería `semopy` usando el siguiente comando si aún no la tienes instalada:

In [3]:
!pip install semopy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 5.5 MB/s eta 0:00:00a 0:00:01
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100.2/100.2 kB 4.6 MB/s eta 0:00:00
  Created wheel for semopy: filename=semopy-2.3.11-py3-none-any.whl size=1659681 sha256=84ea0045fe4964357836e6cb9c9a70ec6ce9d41ca38c91fa264828dc9247c3f7
  Stored in directory: /Users/carloskelly/Library/Caches/pip/wheels/c6/24/8b/be911b059a61f490f38425eb19bf2fed470a5ead97228e8255
Successfully built semopy


## Paso 2: Definir el Modelo SEM
Definimos el modelo SEM en términos de las relaciones entre las variables observadas y latentes. En nuestro caso, modelamos Confianza en la Plataforma y Calidad del Servicio como variables latentes que influyen en la Satisfacción del Cliente.

In [3]:
import pandas as pd
from semopy import Model, semplot
from sklearn.preprocessing import StandardScaler

# Definir el modelo SEM usando la notación estándar de SEM
model_desc = """
# Latent Variables
Confianza =~ Uso + Seguridad + Recomendaciones
Calidad =~ Tiempo_Respuesta + Resolucion_Problemas

# Regressions
Satisfaccion ~ Confianza + Calidad + Ingreso
"""

print(model_desc)


# Latent Variables
Confianza =~ Uso + Seguridad + Recomendaciones
Calidad =~ Tiempo_Respuesta + Resolucion_Problemas

# Regressions
Satisfaccion ~ Confianza + Calidad + Ingreso



## Paso 3: Lectura y normalización de datos
Creamos un conjunto de datos sintéticos que sigue la estructura definida

In [5]:
data = pd.read_csv("data/ecominsights.csv")

In [6]:
data.head()

,Uso,Seguridad,Recomendaciones,Tiempo_Respuesta,Resolucion_Problemas,Ingreso,Satisfaccion
0,4.248357,3.999659,3.550083,4.310602,3.904587,57783.610762,6.092195
1,3.930868,3.762814,2.960261,4.081998,3.587917,44488.142837,4.792502
2,4.323844,4.548032,3.632997,4.208623,3.812221,41818.011166,4.753755
3,4.761515,4.944626,4.080340,4.583536,4.233842,49966.255426,4.832375
4,3.882923,3.876653,3.189026,4.257330,3.841369,48298.153775,4.020023


In [10]:
data.describe()

,Uso,Seguridad,Recomendaciones,Tiempo_Respuesta,Resolucion_Problemas,Ingreso,Satisfaccion
count,300.000000,300.000000,300.000000,300.000000,300.000000,300.000000,300.000000
mean,3.997226,3.990779,3.214219,4.229512,3.824905,50428.230393,4.946916
std,0.492097,0.560252,0.434489,0.302944,0.333527,9845.091535,0.583629
min,2.379366,2.576026,1.870600,3.331123,3.043337,20786.495165,3.152651
25%,3.658377,3.608260,2.897788,4.033161,3.598872,43911.912370,4.595411
50%,4.029610,4.001310,3.212895,4.249547,3.820004,50087.096251,4.899175
75%,4.313329,4.367666,3.505101,4.424827,4.058820,57162.998371,5.289167
max,5.926366,5.695957,4.562171,4.931926,4.713994,81931.075678,6.873606


In [11]:
# Estandarizar los datos
data_std = pd.DataFrame(
    StandardScaler().fit_transform(data),
    columns=data.columns
)

data_std.head()

,Uso,Seguridad,Recomendaciones,Tiempo_Respuesta,Resolucion_Problemas,Ingreso,Satisfaccion
0,0.511182,0.015876,0.774301,0.268120,0.239307,0.748360,1.965620
1,-0.135072,-0.407577,-0.585476,-0.487747,-0.711735,-0.604363,-0.265016
2,0.664837,0.996309,0.965450,-0.069067,-0.038093,-0.876031,-0.331517
3,1.555723,1.705376,1.996755,1.170563,1.228146,-0.047003,-0.196583
4,-0.232664,-0.204045,-0.058082,0.091978,0.049446,-0.216721,-1.590808


In [12]:
data_std.describe()

,Uso,Seguridad,Recomendaciones,Tiempo_Respuesta,Resolucion_Problemas,Ingreso,Satisfaccion
count,3.000000e+02,3.000000e+02,3.000000e+02,3.000000e+02,3.000000e+02,3.000000e+02,3.000000e+02
mean,4.381680e-16,-9.355479e-16,2.131628e-16,-1.521746e-15,6.424491e-16,-5.447494e-16,2.575717e-16
std,1.001671e+00,1.001671e+00,1.001671e+00,1.001671e+00,1.001671e+00,1.001671e+00,1.001671e+00
min,-3.293178e+00,-2.529426e+00,-3.097583e+00,-2.970481e+00,-2.347254e+00,-3.015844e+00,-3.079461e+00
25%,-6.897318e-01,-6.839024e-01,-7.295014e-01,-6.492257e-01,-6.788361e-01,-6.629909e-01,-6.032798e-01
50%,6.591816e-02,1.882928e-02,-3.054223e-03,6.624473e-02,-1.471630e-02,-3.470807e-02,-8.193595e-02
75%,6.434329e-01,6.738343e-01,6.705980e-01,6.457992e-01,7.025102e-01,6.852167e-01,5.873998e-01
max,3.926795e+00,3.048675e+00,3.107570e+00,2.322499e+00,2.670172e+00,3.205199e+00,3.306739e+00


## Paso 4: Ajustar el Modelo SEM a los Datos Sintéticos

In [13]:
mod = Model(model_desc)
res_opt = mod.fit(data_std)
estimates = mod.inspect()

# Imprimir los resultados del ajuste del modelo
print(estimates)

                    lval  op                  rval  Estimate  Std. Err  \
0                    Uso   ~             Confianza  1.000000         -   
1              Seguridad   ~             Confianza  0.857533  0.032977   
2        Recomendaciones   ~             Confianza  0.888696  0.030344   
3       Tiempo_Respuesta   ~               Calidad  1.000000         -   
4   Resolucion_Problemas   ~               Calidad  0.802230  0.088661   
5           Satisfaccion   ~             Confianza  0.395079  0.049052   
6           Satisfaccion   ~               Calidad  0.350719  0.060366   
7           Satisfaccion   ~               Ingreso -0.009977  0.048429   
8                Calidad  ~~               Calidad  0.998294  0.130325   
9              Confianza  ~~             Confianza  1.000057  0.083348   
10             Confianza  ~~               Calidad  0.068095  0.057856   
11       Recomendaciones  ~~       Recomendaciones  0.210046  0.021643   
12  Resolucion_Problemas  ~~  Resoluci

Eliminamos las vriables con p-value > al limite que definamos (0.05 en este caso).

In [14]:
model_desc = """
# Latent Variables
Confianza =~ Uso + Seguridad + Recomendaciones
Calidad =~ Tiempo_Respuesta + Resolucion_Problemas

# Regressions
Satisfaccion ~ Confianza + Calidad
"""

In [15]:
mod = Model(model_desc)
res_opt = mod.fit(data_std)
estimates = mod.inspect()

# Imprimir los resultados del ajuste del modelo
print(estimates)

                    lval  op                  rval  Estimate  Std. Err  \
0                    Uso   ~             Confianza  1.000000         -   
1              Seguridad   ~             Confianza  0.857485  0.032984   
2        Recomendaciones   ~             Confianza  0.888709  0.030355   
3       Tiempo_Respuesta   ~               Calidad  1.000000         -   
4   Resolucion_Problemas   ~               Calidad  0.803084  0.088821   
5           Satisfaccion   ~             Confianza  0.393863   0.04906   
6           Satisfaccion   ~               Calidad  0.350202  0.060357   
7                Calidad  ~~               Calidad  0.997653  0.130379   
8              Confianza  ~~             Confianza  0.999908  0.083338   
9              Confianza  ~~               Calidad  0.068356  0.057863   
10       Recomendaciones  ~~       Recomendaciones  0.210116  0.021652   
11  Resolucion_Problemas  ~~  Resolucion_Problemas  0.356426  0.071731   
12          Satisfaccion  ~~          

In [16]:
estimates

,lval,op,rval,Estimate,Std. Err,z-value,p-value
0,Uso,~,Confianza,1.000000,-,-,-
1,Seguridad,~,Confianza,0.857485,0.032984,25.996642,0.0
2,Recomendaciones,~,Confianza,0.888709,0.030355,29.277026,0.0
3,Tiempo_Respuesta,~,Calidad,1.000000,-,-,-
4,Resolucion_Problemas,~,Calidad,0.803084,0.088821,9.041605,0.0
5,Satisfaccion,~,Confianza,0.393863,0.04906,8.028133,0.0
6,Satisfaccion,~,Calidad,0.350202,0.060357,5.802128,0.0
7,Calidad,~~,Calidad,0.997653,0.130379,7.651956,0.0
8,Confianza,~~,Confianza,0.999908,0.083338,11.998254,0.0
9,Confianza,~~,Calidad,0.068356,0.057863,1.181353,0.237463


In [17]:
confianza_loads = estimates[
    (estimates['rval']=='Confianza') & 
    (estimates['op']=='~') & 
    (estimates['lval']!='Satisfaccion')
][['lval', 'Estimate']].rename(columns={'lval': 'variable', 'Estimate': 'factor_load'})
confianza_loads

,variable,factor_load
0,Uso,1.000000
1,Seguridad,0.857485
2,Recomendaciones,0.888709


In [18]:
calidad_loads = estimates[
    (estimates['rval']=='Calidad') & 
    (estimates['op']=='~') & 
    (estimates['lval']!='Satisfaccion')
][['lval', 'Estimate']].rename(columns={'lval': 'variable', 'Estimate': 'factor_load'})
calidad_loads

,variable,factor_load
3,Tiempo_Respuesta,1.000000
4,Resolucion_Problemas,0.803084


In [19]:
satisfaccion_loads = estimates[
    (estimates['lval']=='Satisfaccion') & 
    (estimates['op']=='~')
][['rval', 'Estimate']].rename(columns={'rval': 'variable', 'Estimate': 'factor_load'})
satisfaccion_loads

,variable,factor_load
5,Confianza,0.393863
6,Calidad,0.350202


In [20]:
def get_factor_load(df, factor_name):
    return df[df['variable']==factor_name]['factor_load'].values[0]

In [21]:
# Función para calcular el valor de la variable latente
def calculate_latent_values(df):
    # Calcular Confianza como la suma ponderada de Uso, Seguridad y Recomendaciones
    df['Confianza'] = (df['Uso'] * get_factor_load(confianza_loads, 'Uso') +
                       df['Seguridad'] * get_factor_load(confianza_loads, 'Seguridad') +
                       df['Recomendaciones'] * get_factor_load(confianza_loads, 'Recomendaciones'))
    
    # Calcular Calidad del Servicio como la suma ponderada de Tiempo de Respuesta y Resolución de Problemas
    df['Calidad'] = (df['Tiempo_Respuesta'] * get_factor_load(calidad_loads, 'Tiempo_Respuesta') +
                     df['Resolucion_Problemas'] * get_factor_load(calidad_loads, 'Resolucion_Problemas'))
    
    # Calcular Satisfacción como la suma ponderada de Confianza y Calidad
    df['Satisfaccion'] = (df['Confianza'] * get_factor_load(satisfaccion_loads, 'Confianza') +
                          df['Calidad'] * get_factor_load(satisfaccion_loads, 'Calidad'))
    
    return df[['Confianza', 'Calidad', 'Satisfaccion']]

In [22]:
# Calcular valores latentes en el dataset
latent_values = calculate_latent_values(data_std)

# Mostrar los primeros valores calculados
print(latent_values.head())


   Confianza   Calidad  Satisfaccion
0   1.212924  0.460304      0.638925
1  -1.004881 -1.059330     -0.766765
2   2.377161 -0.099659      0.901375
3   4.792591  2.156868      2.642963
4  -0.459247  0.131688     -0.134763


In [10]:
g = semplot(mod, "data/pd.png")

### Análisis del Modelo SEM y Resultados

Este análisis está basado en un modelo de ecuaciones estructurales (SEM) que mide la influencia de dos variables latentes (**Confianza en la Plataforma** y **Calidad del Servicio**) sobre la **Satisfacción del Cliente**, además de incorporar el **Ingreso del Hogar** como una variable observada. A continuación, se presenta un resumen de los resultados y las conclusiones derivadas del modelo.

#### Estimaciones de Parámetros

1. **Relaciones entre Variables Latentes y Observadas**:
   - **Uso** (Confianza): Establecido como referencia con un valor de 1.0 (por convención en SEM).
   - **Seguridad** (Confianza): Estima una carga factorial de **0.976**, lo que indica una fuerte relación con la variable latente **Confianza**.
   - **Recomendaciones** (Confianza): Tiene una carga factorial de **0.785**, también indicando una relación significativa con **Confianza**, aunque algo menor que **Seguridad**.
   - **Tiempo de Respuesta** (Calidad): Establecido como referencia con un valor de 1.0.
   - **Resolución de Problemas** (Calidad): Carga factorial de **0.884**, lo que indica una relación muy fuerte con la **Calidad del Servicio**.

2. **Relaciones entre las Variables Latentes y la Satisfacción del Cliente**:
   - **Confianza** → **Satisfacción**: Coeficiente de **0.468**, con una significancia alta (*p-value = 0.0*). Esto indica que a mayor confianza del cliente en la plataforma, mayor es su satisfacción.
   - **Calidad del Servicio** → **Satisfacción**: Coeficiente de **0.676**, también significativo (*p-value = 0.0*), sugiriendo que la calidad del servicio tiene una influencia importante sobre la satisfacción del cliente, incluso mayor que la confianza.
   - **Ingreso del Hogar** → **Satisfacción**: El coeficiente es prácticamente cero (**-0.000**), y no significativo (*p-value = 0.836*). Esto implica que el ingreso del hogar no tiene un impacto relevante en la satisfacción del cliente en este modelo.

3. **Covarianzas**:
   - **Confianza** y **Calidad del Servicio** están correlacionadas positivamente (**0.010**), aunque esta relación no es estadísticamente significativa (*p-value = 0.237*).
   - La varianza residual de **Satisfacción del Cliente** es **0.239**, lo que indica que el modelo captura una buena parte de la varianza en satisfacción, pero aún hay factores no explicados que influyen.

#### Métricas de Ajuste del Modelo

El ajuste del modelo fue satisfactorio, con un valor objetivo de **0.046** después de 30 iteraciones utilizando el método de optimización **SLSQP**. El modelo se ajustó correctamente a los datos, lo que se refleja en el éxito del proceso de optimización.

#### Interpretación General

- **Confianza en la Plataforma** y **Calidad del Servicio** son dos factores clave que explican la **Satisfacción del Cliente** en la plataforma de e-commerce. Ambos factores tienen un impacto significativo y positivo, siendo la **Calidad del Servicio** el factor más importante en términos de influencia sobre la satisfacción.
  
- **Ingreso del Hogar** no parece tener un impacto directo en la satisfacción del cliente, lo que sugiere que la experiencia del usuario no depende significativamente de su nivel socioeconómico en este caso.

- Las relaciones entre las variables latentes y sus indicadores observados son fuertes y significativas, lo que valida la construcción teórica de las variables latentes de **Confianza** y **Calidad del Servicio**.

#### Conclusiones

Este modelo de ecuaciones estructurales ha demostrado ser útil para entender cómo factores complejos como la confianza y la calidad del servicio impactan en la satisfacción del cliente. Las empresas pueden enfocarse en mejorar estos dos aspectos, especialmente en el servicio, para aumentar la satisfacción general de los usuarios. Además, las intervenciones en mejorar la percepción de seguridad y la resolución de problemas pueden ser prioritarias, dado su peso en los constructos latentes.

Este análisis proporciona una base sólida para la toma de decisiones basada en datos dentro de la empresa, dirigiendo los esfuerzos hacia las áreas que mayor impacto tienen en la satisfacción del cliente.
